les/access/context.go

```diff
@@ -0,0 +1,78 @@
+// Copyright 2015 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+// Package access provides a layer to handle local blockchain database and
+// on-demand network retrieval
+package access
+
+import (
+	"sync/atomic"
+	"time"
+)
+
+// Context for ODR requests, can be timed out or cancelled manually
+type OdrContext struct {
+	cancel, cancelOrTimeout chan struct{}
+	id                      *OdrChannelID
+	cancelled               int32
+}
+
+var NullCtx = (*OdrContext)(nil) // used when creating states
+var NoOdr = (*OdrContext)(nil)   // used for individual requests
```

Comment:
In the tutorials of the built-in go Context type it is explicitly advised that you should not use nil, exactly because of readability issues. And I agree with this idea, during development it was extremely useful to be able to search for places where I specified null contexts. Specifying "don't use ODR, fail instead when it's not in the DB" is a very explicit statement that should not be considered default and made invisible.
Btw OdrContext is not a descendant of Context because it was easier to implement my own context for my own purpose but the same logic applies.